plugin/utils/compiler_builtins.py

```diff
@@ -0,0 +1,197 @@
+        "__STDC_HOSTED__",
+        "__REGISTER_PREFIX__",
+        "__STDC__"
+    ]
+
+    def __init__(self, args):
+        """
+        Create an object holding the built-in flags of a compiler.
+
+        This constructs a new object which holds the built-in flags
+        used by a compiler. The `args` is the call to the compiler; either
+        a string or a list of strings. If a list of strings is provided, it
+        is interpreted as the call of a compiler (i.e. the first entry
+        is the compiler to call and everything else are arguments to the
+        compiler). If a single string is given, it is parsed into a string
+        list first.
+        """
+        from shlex import split
+        super().__init__()
+        self._defines = list()
+        self._include_paths = list()
+        if isinstance(args, str):
+            # Parse arguments into list of strings first
+            args = split(args)
+        # Guess the compiler and standard:
+        (compiler, std) = self._guess_compiler(args)
+        if compiler is not None:
+            # Guess the language (we need to pass it to the compiler
+            # explicitly):
+            language = self._guess_language(compiler)
+            # Get defines and include paths from the compier:
+            cfg = (compiler, std, language)
+            self._debug("Getting default flags for {}".format(cfg))
+            if cfg in CompilerBuiltIns.__cache:
+                self._debug("Reusing flags from cache")
+                (defines, includes) = CompilerBuiltIns.__cache[cfg]
+            else:
+                self._debug("Querying compiler for defaults")
+                defines = self._get_default_flags(compiler, std, language)
+                includes = self._get_default_include_paths(
+                    compiler, std, language)
+                CompilerBuiltIns.__cache[cfg] = (defines, includes)
+            self._defines = defines
+            self._include_paths = includes
+
+
+    @property
+    def defines(self):
+        """The built-in defines provided by the compiler."""
+        return self._defines
+
+    @property
+    def include_paths(self):
+        """The list of built-in include paths used by the compiler."""
+        return self._include_paths
+
+    @property
+    def flags(self):
+        """
+        The list of built-in flags.
+
+        This property holds the combined list of built-in defines and
+        include paths of the compiler.
+        """
+        return self._defines + self._include_paths
+
+    def _guess_compiler(self, args):
+        compiler = None
+        std = None
+        if len(args) > 0:
+            compiler = args[0]
+        else:
+            self._debug("Got empty command line - cannot extract compiler")
+        if len(args) > 1:
+            for arg in args[1:]:
+                if arg.startswith("-std="):
+                    std = arg[5:]
+        return (compiler, std)
+
+    def _guess_language(self, compiler):
```

Comment:
I have two questions to this.
- This will not work with objective c, right? We should make it clear in the settings, I guess. I am not the one who added the support for it in the first place, but I don't want to break it.
- Can we unify this function with [this](https://github.com/niosus/EasyClangComplete/blob/master/plugin/tools.py#L522) function?

Reply:
Right, currently, this only distinguishes C++ from the "rest" (where we assume that rest refers to C). I think we can try to pimp this utility a bit to also make it work with other languages, I just never works with Objective C/C++ before and hence did not care in this first shot yet.

Regarding the existing method you mentioned: I already stumbled upon it, however, I chose not to make use of it as it requires a `view` object and I don't see how to get in in the place where I do the guessing thing. If you see a way how we can properly pass it through, please let me know and I'll try to implement it.

Reply:
I did not work with objective c either, but we should at least make it clear to the user that it *is not supposed* to work as opposed to just silently failing or generating garbage. I guess a clean fallback to using `common_flags` would be appropriate if the language is different from c or c++. This way if somebody decides to use it with objective c we can ask them to adapt the code to account for the missing functionality :)

Reply:
As for the function and what it does. Currently, I feel like this is not a very scalable solution. For example, `clang++` actually symlinks to `clang`: 
```
$ ls -la | grep clang             
-rwxr-xr-x 1 root root 22938632 Jul 12  2016 clang
lrwxrwxrwx 1 root root        5 Jul 12  2016 clang++ -> clang
```
Am I correct that if the user would use `clang` this would report `c`?

Reply:
Regarding your first comment: Yes, the function definitely needs some love - silently falling back to C should probably not be used by default.

Regarding your second comment: It is normal that `clang++` is sym-linked to `clang` - this is a normal approach also for other compilers. The `clang` binary internally checks by which name it was called and either activates C or C++ mode (at least this is my understanding).

The background the method is written like it that is, that it is called on a compiler command line like

```
path/to/c++ -I/some/include/path -DSOME_DEFINE my_file.cpp
```
To get the correct defines and include paths for code completion, we must - when asking the compiler - explicitly tell it for which language, i.e. we must pass it the `-x SOMELANGUAGE` flag - "just" using the correctly symlinked version will for whatever reasons not work as expected :-(

Currently, the method tries to derive the language only from the compiler name. This is not quite smart (e.g. if the command line in a `compile_commands.json` contains a `-x` flag, we should probably use that instead of looking at the compiler name), but it works for the beginning.

I think your concerns about what a user might select would only kick in if we do not work on a given call to a compiler (as we get it from CMake and a compile database) but instead would have to guess from other pieces of information. In these cases, e.g. the selected language of the current view or the extension of a file should rather be used to decide.

Reply:
I just looked into the code once more and also into how e.g. a typical `compile_commands.json` generated by CMake looks like. My observation for now is:

* By default, CMake does not include a `-x` flag from which we can guess the language from.
* On my system, I checked the `gcc-objc` and `gcc-objc++` packages: They do *not* seem to install additional binaries; so I guess if one would like to use them CMake and other tools would have to explicitly ask for a language using the `-x` flag.
* In addition, I compared the list of flags that we get e.g. from GCC when asking for different languages (C, C++ and ObjC). The lists seems vastly identical, with some specific defines usually added to indicate a particular language's capabilities. Hence, I would assume that the flags for C are mostly a subset of all other supported languages.

Hence, I would indeed suggest the following:

1. When guessing the language, we check first if we find an explicit `-x` flag and use it regardless what else is in the command line.
2. In any other case, we try to guess the name from the compiler. If the compiler name ends with `c++` or `++`, we assume the file in question is compiled with C++, otherwise, we assume C. That should be relatively save, because the flags used for C form (mostly) a subset of any other languages' flags.

If we would not allow point 2., we probably could not provide this feature for C sources altogether (at least I currently don't see how we could).

What do you think?

Reply:
I think this is fine for beginning. I am still thinking more in the direction of deriving the correct language from a view, even if it means that we need to propagate it down into this class (or just get language beforewise). The good part about using the view-dependent detection is that there is a closed loop to the user. I mean, the user sees which syntax is being used for a file, which makes it easy for us to avoid ambiguities. 

But again, I believe your current suggestion is a good first step. Just probably leave a TODO in the code to not forget that there can be an alternative.